### PR TITLE
When thread exits, close sampling ring buffer and file descriptor

### PR DESCRIPTION
**Pull request by dpallotti**
Unfortunately this requires some extra bookkeeping.

### CODE REVIEW

File: OrbitCore/LinuxEventTracer.cpp [hunk 1]
```diff
@@ -211,17 +217,18 @@ void LinuxEventTracerThread::Run(
 
             if (fork.PID() == pid_) {
               // A new thread of the sampled process was spawned.
-              new_thread = fork.TID();
-              // Do not add a new ring buffer to ring_buffers here as we are
-              // already iterating over ring_buffers.
+              thread_forked = fork.TID();
+              // Do not add a new ring buffer to fds_to_ring_buffer here as we
+              // are already iterating over fds_to_ring_buffer.
```
Comment (dimitry-) [inline OrbitCore/LinuxEventTracer.cpp]: This is hard to follow. Maybe have a scheduled_actions list, add a AddThread/RemoveThread actions here and have them invoked after the loop is complete?

File: OrbitCore/LinuxEventTracer.cpp [hunk 2]
```diff
@@ -115,27 +119,29 @@ void LinuxEventTracerThread::Run(
...
+
+      bool is_uprobes = uprobe_fds_to_function.count(fd) > 0;
+      bool is_uretprobes = uretprobe_fds_to_function.count(fd) > 0;
 
       uint32_t read_from_this_buffer = 0;
       // Read up to ROUND_ROBIN_BATCH_SIZE (5) new events
       while (ring_buffer.HasNewData() &&
              read_from_this_buffer < ROUND_ROBIN_BATCH_SIZE) {
-        if (*a_ExitRequested || new_thread >= 0) {
+        if (*a_ExitRequested || thread_forked >= 0 || thread_exited >= 0) {
```
Comment (dimitry-) [inline OrbitCore/LinuxEventTracer.cpp]: why do we want to break when thread_forked/exited is not -1?

also replace >= 0 with != -1

File: OrbitCore/LinuxEventTracer.cpp [hunk 3]
```diff
@@ -34,20 +34,26 @@ void LinuxEventTracerThread::Run(
...
+        int32_t context_switch_fd =
+            LinuxPerfUtils::context_switch_open(-1, cpu);
+        fds_to_ring_buffer.emplace(context_switch_fd,
+                                   LinuxPerfRingBuffer{context_switch_fd});
       }
     } else {
       // perf_event_open for all cpus and the PID to keep track of process
       // spawning.
-      fds.push_back(LinuxPerfUtils::context_switch_open(pid_, -1));
+      int32_t context_switch_fd = LinuxPerfUtils::context_switch_open(pid_, -1);
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/LinuxEventTracer.cpp]: Done (slightly different names).

File: OrbitCore/LinuxEventTracer.cpp (hunk 1 above)
Comment (dpallotti) [reply OrbitCore/LinuxEventTracer.cpp]: Done as discussed offline.

File: OrbitCore/LinuxEventTracer.cpp (hunk 2 above)
Comment (dpallotti) [reply OrbitCore/LinuxEventTracer.cpp]: Done.
